File: modules/ble-peripheral/ios/BlePeripheralModule.swift
import CoreBluetooth
import ExpoModulesCore

/**
 아이폰이 블루투스로 자기를 알린다.

 **자주 쓰는 라이브러리는 찾는 쪽만 한다.** 알리는 쪽은 마땅한 것이
 없어서 직접 만든다. 핫스팟을 못 쓸 때 — 항공사가 개인 핫스팟을
 금지할 때 — 글로라도 대화하려면 이게 필요하다.

 Wi-Fi 와 역할이 반대다. 아이폰이 알리고 안드로이드가 찾는다.
 각 운영체제가 잘하는 쪽에 맞췄다. iOS 는 화면이 꺼져도 제한적으로
 알림을 이어가고, 안드로이드는 찾는 동작에 제약이 적다.

 (docs/04-transport-spec.md 4장 · T20)
 */

private let serviceUUID = CBUUID(string: "F7D5061F-298D-46DB-BE86-D1E0C757AB23")
private let inboxUUID = CBUUID(string: "0F52C6B6-FDDA-4E0C-886A-0BB1CA261494")
private let outboxUUID = CBUUID(string: "8966097A-CE16-4B4B-99C4-5DD2409FC9F7")
private let statusUUID = CBUUID(string: "F9F2387C-9014-4C34-AB13-EC01330DB314")
private let controlUUID = CBUUID(string: "57998665-F536-4E22-BBE8-0302580269C6")

public class BlePeripheralModule: Module {
  private var manager: CBPeripheralManager?
  private var delegate: PeripheralDelegate?

  public func definition() -> ModuleDefinition {
    Name("BlePeripheral")

    Events("onReceive", "onStateChange", "onSubscribe")

    /// 이 기기에서 알릴 수 있나. **없을 수 있다는 전제로 쓴다.**
    Function("isSupported") { () -> Bool in
      return true
    }

    AsyncFunction("start") { (localName: String, promise: Promise) in
      if self.manager != nil {
        promise.resolve(true)
        return
      }

      let handler = PeripheralDelegate(
        localName: localName,
        emit: { [weak self] name, body in
          self?.sendEvent(name, body)
        }
      )

      self.delegate = handler
      self.manager = CBPeripheralManager(delegate: handler, queue: nil)
      handler.manager = self.manager

      promise.resolve(true)
    }

    AsyncFunction("stop") { (promise: Promise) in
      self.delegate?.teardown()
      self.manager = nil
      self.delegate = nil
      promise.resolve(true)
    }

    /// 조각 하나를 상대에게 보낸다.
    ///
    /// 보낼 자리가 없으면 `false` 를 돌려준다. **거기서 멈추고
    /// 기다려야 한다.** 계속 밀어 넣으면 조용히 버려진다.
    AsyncFunction("send") { (base64: String, promise: Promise) in
      guard let handler = self.delegate else {
        promise.resolve(false)
        return
      }
      promise.resolve(handler.send(base64: base64))
    }

    Function("isAdvertising") { () -> Bool in
      return self.manager?.isAdvertising ?? false
    }
  }
}

private class PeripheralDelegate: NSObject, CBPeripheralManagerDelegate {
  weak var manager: CBPeripheralManager?

  private let localName: String
  private let emit: (String, [String: Any]) -> Void

  private var outbox: CBMutableCharacteristic?
  private var subscribers: [CBCentral] = []

  /// 보낼 자리가 없어 밀려난 것들. 자리가 나면 여기부터 보낸다.
  private var backlog: [Data] = []

  init(localName: String, emit: @escaping (String, [String: Any]) -> Void) {
    self.localName = localName
    self.emit = emit
  }

  func teardown() {
    manager?.stopAdvertising()
    manager?.removeAllServices()
    subscribers.removeAll()
    backlog.removeAll()
  }

  func send(base64: String) -> Bool {
    guard
      let data = Data(base64Encoded: base64),
      let outbox = outbox,
      !subscribers.isEmpty
    else {
      return false
    }

    // 밀린 것이 있으면 순서를 지킨다. 새것을 먼저 보내면 말이 뒤엉킨다.
    if !backlog.isEmpty {
      backlog.append(data)
      return flush()
    }

    let sent = manager?.updateValue(
      data,
      for: outbox,
      onSubscribedCentrals: subscribers
    ) ?? false

    if !sent { backlog.append(data) }
    return sent
  }

  /// 자리가 났을 때 밀린 것을 내보낸다
  @discardableResult
  private func flush() -> Bool {
    guard let outbox = outbox, !subscribers.isEmpty else { return false }

    while let next = backlog.first {
      let sent = manager?.updateValue(
        next,
        for: outbox,
        onSubscribedCentrals: subscribers
      ) ?? false

      if !sent { return false }
      backlog.removeFirst()
    }

    return true
  }

  func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    emit("onStateChange", ["state": describe(peripheral.state)])

    guard peripheral.state == .poweredOn else { return }
    publish(on: peripheral)
  }

  private func publish(on peripheral: CBPeripheralManager) {
    let inbox = CBMutableCharacteristic(
      type: inboxUUID,
      properties: [.writeWithoutResponse, .write],
      value: nil,
      permissions: [.writeable]
    )

    let out = CBMutableCharacteristic(
      type: outboxUUID,
      properties: [.notify],
      value: nil,
      permissions: [.readable]
    )

    let status = CBMutableCharacteristic(
      type: statusUUID,
      properties: [.read, .notify],
      value: nil,
      permissions: [.readable]
    )

    let control = CBMutableCharacteristic(
      type: controlUUID,
      properties: [.write],
      value: nil,
      permissions: [.writeable]
    )

    let service = CBMutableService(type: serviceUUID, primary: true)
    service.characteristics = [inbox, out, status, control]

    outbox = out

    peripheral.removeAllServices()
    peripheral.add(service)
  }

  func peripheralManager(
    _ peripheral: CBPeripheralManager,
    didAdd service: CBService,
    error: Error?
  ) {
    guard error == nil else {
      emit("onStateChange", ["state": "failed"])
      return
    }

    peripheral.startAdvertising([
      CBAdvertisementDataServiceUUIDsKey: [serviceUUID],
      CBAdvertisementDataLocalNameKey: localName,
    ])
  }

  func peripheralManager(
    _ peripheral: CBPeripheralManager,
    central: CBCentral,
    didSubscribeTo characteristic: CBCharacteristic
  ) {
    guard characteristic.uuid == outboxUUID else { return }

    if !subscribers.contains(where: { $0.identifier == central.identifier }) {
      subscribers.append(central)
    }

    emit("onSubscribe", ["subscribed": true, "mtu": central.maximumUpdateValueLength])
  }

  func peripheralManager(
    _ peripheral: CBPeripheralManager,
    central: CBCentral,
    didUnsubscribeFrom characteristic: CBCharacteristic
  ) {
    subscribers.removeAll { $0.identifier == central.identifier }
    // 아무도 안 듣는데 쌓아둬도 갈 곳이 없다
    if subscribers.isEmpty { backlog.removeAll() }

    emit("onSubscribe", ["subscribed": false, "mtu": 0])
  }

  func peripheralManager(
    _ peripheral: CBPeripheralManager,
    didReceiveWrite requests: [CBATTRequest]
  ) {
    for request in requests {
      guard let value = request.value else { continue }
      emit("onReceive", ["data": value.base64EncodedString()])
    }

    // 답을 요구하는 쓰기에는 답해야 한다. 안 그러면 상대가 멈춘다.
    if let first = requests.first {
      peripheral.respond(to: first, withResult: .success)
    }
  }

  func peripheralManagerIsReady(toUpdateSubscribers peripheral: CBPeripheralManager) {
    flush()
  }

  private func describe(_ state: CBManagerState) -> String {
    switch state {
    case .poweredOn: return "on"
    case .poweredOff: return "off"
    case .unauthorized: return "denied"
    case .unsupported: return "unsupported"
    default: return "unknown"
    }
  }
}

File: modules/ble-peripheral/android/src/main/java/com/ybbarng/skywalkie/bleperipheral/BlePeripheralModule.kt
package com.ybbarng.skywalkie.bleperipheral

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothGattServer
import android.bluetooth.BluetoothGattServerCallback
import android.bluetooth.BluetoothGattService
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.BluetoothLeAdvertiser
import android.content.Context
import android.os.Build
import android.os.ParcelUuid
import android.util.Base64
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID

/**
 * 안드로이드가 블루투스로 자기를 알린다.
 *
 * **아이폰 쪽과 똑같은 이름(`BlePeripheral`)을 쓴다.** 그래야 JS 가
 * 어느 폰인지 신경 쓰지 않고 같은 코드를 부른다.
 *
 * ## 왜 안드로이드가 알리는 쪽이어야 하나
 *
 * iOS 는 앱이 뒤로 가면 알림에 담긴 서비스 번호를 **애플 기기만 읽는
 * 자리로 옮긴다.** 그래서 아이폰이 알리는 쪽이면 아이폰 앱이 앞에 떠
 * 있을 때만 상대가 찾을 수 있다. 여자친구가 영화를 보는 동안 연결이
 * 끊기면 **앱을 다시 열기 전까지 복구가 안 된다.**
 *
 * 반대로 안드로이드는 앞쪽 알림을 띄운 채 계속 알릴 수 있고, iOS 는
 * 뒤에 있어도 정해둔 서비스 번호로 찾을 수 있다. 그래서 이쪽이 맞다.
 *
 * ## 안드로이드끼리도 이어진다
 *
 * 이 모듈이 생기면서 **안드로이드 두 대로도 시험할 수 있다.** 전에는
 * 아이폰만 알릴 수 있어서 둘 다 찾기만 하다 영영 못 만났다.
 *
 * ## 없으면 없는 대로 둔다
 *
 * 알리는 장치가 없는 기기가 있다. 그때는 "못 한다" 고만 답하고
 * **여기서 앱이 죽으면 안 된다.** 덤 하나 때문에 대화를 통째로 못
 * 하게 되는 것이 이 앱에서 가장 나쁜 일이다.
 *
 * (docs/04-transport-spec.md 4장 · T20)
 */

private val SERVICE_UUID: UUID = UUID.fromString("F7D5061F-298D-46DB-BE86-D1E0C757AB23")
private val INBOX_UUID: UUID = UUID.fromString("0F52C6B6-FDDA-4E0C-886A-0BB1CA261494")
private val OUTBOX_UUID: UUID = UUID.fromString("8966097A-CE16-4B4B-99C4-5DD2409FC9F7")
private val STATUS_UUID: UUID = UUID.fromString("F9F2387C-9014-4C34-AB13-EC01330DB314")
private val CONTROL_UUID: UUID = UUID.fromString("57998665-F536-4E22-BBE8-0302580269C6")

/**
 * 듣겠다고 표시하는 자리.
 *
 * 블루투스가 정해둔 번호다. 우리가 고른 것이 아니다. 찾는 쪽이 여기에
 * 쓰면 "알림을 받겠다" 는 뜻이 된다.
 */
private val CCCD_UUID: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

/** 협상 전 기본값. 3바이트는 블루투스가 가져간다 */
private const val DEFAULT_MTU = 23
private const val ATT_OVERHEAD = 3

class BlePeripheralModule : Module() {
  private var adapter: BluetoothAdapter? = null
  private var advertiser: BluetoothLeAdvertiser? = null
  private var server: BluetoothGattServer? = null
  private var outbox: BluetoothGattCharacteristic? = null

  /** 듣고 있는 상대들 */
  private val listeners = mutableListOf<BluetoothDevice>()

  /** 보낼 자리가 없어 밀려난 것들. 자리가 나면 여기부터 보낸다 */
  private val backlog = ArrayDeque<ByteArray>()

  private var mtu = DEFAULT_MTU
  private var advertising = false

  /**
   * 알리기를 시작해달라고 한 뒤 답을 기다리는 쪽.
   *
   * **`startAdvertising` 을 부른다고 알려지는 게 아니다.** 되는지 안
   * 되는지는 여기 콜백으로 나중에 온다. 이걸 안 기다리면 실패했는데도
   * 열린 것으로 치고, 아무도 우리를 못 보는 채로 기다리게 된다.
   */
  private var pending: Promise? = null

  private fun settle(started: Boolean) {
    val waiting = pending ?: return
    pending = null
    waiting.resolve(started)
  }

  private val advertiseCallback = object : AdvertiseCallback() {
    override fun onStartSuccess(settingsInEffect: AdvertiseSettings) {
      advertising = true
      emitState("on")
      settle(true)
    }

    override fun onStartFailure(errorCode: Int) {
      advertising = false
      // **여기서 예외를 던지지 않는다.** 알리기가 안 되는 것뿐이고
      // 다른 길은 아직 살아 있다.
      emitState("failed")
      settle(false)
    }
  }

  private val serverCallback = object : BluetoothGattServerCallback() {
    override fun onConnectionStateChange(device: BluetoothDevice, status: Int, newState: Int) {
      if (newState == BluetoothProfile.STATE_DISCONNECTED) {
        listeners.removeAll { it.address == device.address }
        if (listeners.isEmpty()) {
          backlog.clear()
          mtu = DEFAULT_MTU
        }
        emitSubscribe(false, 0)
      }
    }

    override fun onMtuChanged(device: BluetoothDevice, size: Int) {
      mtu = size
      // 이미 듣고 있는 상대라면 새 크기를 알려준다. 넘겨 보내면
      // 조용히 잘린다.
      if (listeners.any { it.address == device.address }) {
        emitSubscribe(true, usableBytes())
      }
    }

    override fun onCharacteristicWriteRequest(
      device: BluetoothDevice,
      requestId: Int,
      characteristic: BluetoothGattCharacteristic,
      preparedWrite: Boolean,
      responseNeeded: Boolean,
      offset: Int,
      value: ByteArray?,
    ) {
      if (characteristic.uuid == INBOX_UUID && value != null) {
        emit("onReceive", mapOf("data" to Base64.encodeToString(value, Base64.NO_WRAP)))
      }

      // 답을 요구하는 쓰기에는 답해야 한다. 안 그러면 상대가 멈춘다.
      if (responseNeeded) {
        server?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, null)
      }
    }

    override fun onDescriptorWriteRequest(
      device: BluetoothDevice,
      requestId: Int,
      descriptor: BluetoothGattDescriptor,
      preparedWrite: Boolean,
      responseNeeded: Boolean,
      offset: Int,
      value: ByteArray?,
    ) {
      if (descriptor.uuid == CCCD_UUID && descriptor.characteristic.uuid == OUTBOX_UUID) {
        val wantsIt = value != null &&
          value.contentEquals(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE)

        if (wantsIt) {
          if (listeners.none { it.address == device.address }) listeners.add(device)
          emitSubscribe(true, usableBytes())
        } else {
          listeners.removeAll { it.address == device.address }
          if (listeners.isEmpty()) backlog.clear()
          emitSubscribe(false, 0)
        }
      }

      if (responseNeeded) {
        server?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, null)
      }
    }

    override fun onNotificationSent(device: BluetoothDevice, status: Int) {
      // 자리가 났다. 밀린 것을 마저 보낸다.
      if (status == BluetoothGatt.GATT_SUCCESS) flush()
    }
  }

  override fun definition() = ModuleDefinition {
    Name("BlePeripheral")

    Events("onReceive", "onStateChange", "onSubscribe")

    /** 이 기기에서 알릴 수 있나. **없을 수 있다는 전제로 쓴다.** */
    Function("isSupported") {
      runCatching { bluetoothAdapter()?.isMultipleAdvertisementSupported == true }
        .getOrDefault(false)
    }

    AsyncFunction("start") { localName: String, promise: Promise ->
      if (advertising) {
        promise.resolve(true)
        return@AsyncFunction
      }

      pending = promise

      val asked = runCatching { startAdvertising(localName) }.getOrElse {
        emitState("failed")
        false
      }

      // 시작조차 못 했으면 콜백이 안 온다. 여기서 끝낸다.
      if (!asked) settle(false)
    }

    AsyncFunction("stop") {
      runCatching { stopAdvertising() }.getOrDefault(false)
    }

    /**
     * 조각 하나를 보낸다.
     *
     * 보낼 자리가 없으면 `false` 를 돌려준다. **거기서 멈추고 기다려야
     * 한다.** 계속 밀어 넣으면 조용히 버려진다.
     */
    AsyncFunction("send") { base64: String ->
      runCatching { sendChunk(base64) }.getOrDefault(false)
    }

    Function("isAdvertising") { advertising }
  }

  private fun bluetoothAdapter(): BluetoothAdapter? {
    if (adapter != null) return adapter

    val context = appContext.reactContext ?: return null
    val manager = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
    adapter = manager?.adapter
    return adapter
  }

  private fun startAdvertising(localName: String): Boolean {
    if (advertising) return true

    val context = appContext.reactContext ?: return false
    val bluetooth = bluetoothAdapter() ?: run {
      emitState("unsupported")
      return false
    }

    if (!bluetooth.isEnabled) {
      emitState("off")
      return false
    }

    val manager = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
      ?: return false

    // **권한이 없으면 여기서 예외가 난다.** 잡아서 "못 한다" 로
    // 돌려준다. 앱이 죽으면 다른 길까지 막힌다.
    val opened = manager.openGattServer(context, serverCallback) ?: run {
      emitState("denied")
      return false
    }
    server = opened

    opened.addService(buildService())

    val ble = bluetooth.bluetoothLeAdvertiser ?: run {
      emitState("unsupported")
      return false
    }
    advertiser = ble

    // **이름을 알림에 담지 않는다.** 알림 한 칸이 31바이트뿐인데
    // 서비스 번호(16바이트)만으로도 꽉 찬다. 이름까지 넣으면 알리기가
    // 통째로 실패한다. 이름은 이어진 뒤 인사 봉투로 주고받는다.
    //
    // **폰의 블루투스 이름도 건드리지 않는다.** 남의 이름을 바꾸는
    // 일이고, 권한에 따라 터지기도 한다.

    val settings = AdvertiseSettings.Builder()
      .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_BALANCED)
      .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_HIGH)
      .setConnectable(true)
      .setTimeout(0)
      .build()

    val data = AdvertiseData.Builder()
      .addServiceUuid(ParcelUuid(SERVICE_UUID))
      .setIncludeDeviceName(false)
      .setIncludeTxPowerLevel(false)
      .build()

    ble.startAdvertising(settings, data, advertiseCallback)
    return true
  }

  private fun buildService(): BluetoothGattService {
    val service = BluetoothGattService(SERVICE_UUID, BluetoothGattService.SERVICE_TYPE_PRIMARY)

    val inbox = BluetoothGattCharacteristic(
      INBOX_UUID,
      BluetoothGattCharacteristic.PROPERTY_WRITE or
        BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE,
      BluetoothGattCharacteristic.PERMISSION_WRITE,
    )

    val out = BluetoothGattCharacteristic(
      OUTBOX_UUID,
      BluetoothGattCharacteristic.PROPERTY_NOTIFY,
      BluetoothGattCharacteristic.PERMISSION_READ,
    )
    out.addDescriptor(
      BluetoothGattDescriptor(
        CCCD_UUID,
        BluetoothGattDescriptor.PERMISSION_READ or BluetoothGattDescriptor.PERMISSION_WRITE,
      ),
    )
    outbox = out

    val status = BluetoothGattCharacteristic(
      STATUS_UUID,
      BluetoothGattCharacteristic.PROPERTY_READ or
        BluetoothGattCharacteristic.PROPERTY_NOTIFY,
      BluetoothGattCharacteristic.PERMISSION_READ,
    )
    status.addDescriptor(
      BluetoothGattDescriptor(
        CCCD_UUID,
        BluetoothGattDescriptor.PERMISSION_READ or BluetoothGattDescriptor.PERMISSION_WRITE,
      ),
    )

    val control = BluetoothGattCharacteristic(
      CONTROL_UUID,
      BluetoothGattCharacteristic.PROPERTY_WRITE,
      BluetoothGattCharacteristic.PERMISSION_WRITE,
    )

    service.addCharacteristic(inbox)
    service.addCharacteristic(out)
    service.addCharacteristic(status)
    service.addCharacteristic(control)
    return service
  }

  private fun stopAdvertising(): Boolean {
    runCatching { advertiser?.stopAdvertising(advertiseCallback) }
    runCatching { server?.close() }

    advertiser = null
    server = null
    outbox = null
    listeners.clear()
    backlog.clear()
    mtu = DEFAULT_MTU
    advertising = false
    return true
  }

  private fun sendChunk(base64: String): Boolean {
    val out = outbox ?: return false
    if (listeners.isEmpty()) return false

    val data = runCatching { Base64.decode(base64, Base64.NO_WRAP) }.getOrNull() ?: return false

    // 밀린 것이 있으면 순서를 지킨다. 새것을 먼저 보내면 말이 뒤엉킨다.
    if (backlog.isNotEmpty()) {
      backlog.addLast(data)
      return flush()
    }

    return notify(out, data)
  }

  /** 자리가 났을 때 밀린 것을 내보낸다 */
  private fun flush(): Boolean {
    val out = outbox ?: return false
    if (listeners.isEmpty()) return false

    while (backlog.isNotEmpty()) {
      val next = backlog.first()
      if (!notify(out, next)) return false
      backlog.removeFirst()
    }

    return true
  }

  private fun notify(out: BluetoothGattCharacteristic, data: ByteArray): Boolean {
    val gatt = server ?: return false

    for (device in listeners) {
      val sent = runCatching {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
          gatt.notifyCharacteristicChanged(device, out, false, data) ==
            BluetoothGatt.GATT_SUCCESS
        } else {
          @Suppress("DEPRECATION")
          out.value = data
          @Suppress("DEPRECATION")
          gatt.notifyCharacteristicChanged(device, out, false)
        }
      }.getOrDefault(false)

      if (!sent) {
        // 자리가 없다. 밀어두고 `onNotificationSent` 를 기다린다.
        if (backlog.firstOrNull() !== data) backlog.addFirst(data)
        return false
      }
    }

    return true
  }

  /** 한 번에 실어 보낼 수 있는 바이트 */
  private fun usableBytes(): Int = mtu - ATT_OVERHEAD

  private fun emitState(state: String) {
    emit("onStateChange", mapOf("state" to state))
  }

  private fun emitSubscribe(subscribed: Boolean, size: Int) {
    emit("onSubscribe", mapOf("subscribed" to subscribed, "mtu" to size))
  }

  private fun emit(name: String, body: Map<String, Any>) {
    // 듣는 쪽이 없을 수 있다. 그것 때문에 죽지 않는다.
    runCatching { sendEvent(name, body) }
  }
}
